scipy/stats/_AB_SW.pyx

```diff
@@ -0,0 +1,501 @@
+# cython: language_level=3
+# cython: boundscheck=False
+# cython: initializedcheck=False
+# cython: nonecheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# cython: cpow=True
+
+from libc.math cimport exp, sqrt, abs, log, acos
+import numpy as np
+cimport numpy as cnp
+cnp.import_array()
+
+
+def gscale(int test, int other):
+    """
+    Cython translation for the FORTRAN 77 code given in:
+
+    Dinneen, L. C. and Blakesley, B. C., "Algorithm AS 93: A Generator for the
+    Null Distribution of the Ansari-Bradley W Statistic", Applied Statistics,
+    25(1), 1976, :doi:`10.2307/2346534`
+    """
+    cdef int m = min(test, other), n = max(test, other)
+    cdef int astart = ((test + 1) // 2) * (1 + (test // 2))
+    cdef int LL = (test * other) // 2 + 1
+    cdef int ind, n2b1, n2b2, loop_m = 3, part_no, ks
+    cdef int len1 = 0, len2 = 0, len3 = 0  # Nonzero entry lengths
+
+    cdef cnp.ndarray a1 = cnp.PyArray_ZEROS(1, [LL], cnp.NPY_FLOAT32, 0)
+    cdef cnp.ndarray a2 = cnp.PyArray_ZEROS(1, [LL], cnp.NPY_FLOAT32, 0)
+    cdef cnp.ndarray a3 = cnp.PyArray_ZEROS(1, [LL], cnp.NPY_FLOAT32, 0)
+    cdef float[::1] a1v = a1  # memview of a1
+    cdef float[::1] a2v = a2  # memview of a2
+    cdef float[::1] a3v = a3  # memview of a3
```

Comment:
Memoryiews are faster with element access and many for loops are done over views. But they can't do numpy ops like `@` or `+=`-like infix ops. So we create the numpy arrays for that kind of ops and put memoryview on top of their memory to do array access. It's really not very nice for my taste, but performance goes up somehow with memoryviews I guess via skipping NumPy overhead. 

Cython 3 is said to fix some of these things but can't say anything before trying.

Reply:
Okay, if it helped let's keep it. It'd be useful if you pushed a commit with a comment to explain why the code is like this - because it's the kind of thing that the next person who comes along may try to clean up otherwise.

Reply:
Yes let's do it when it's ready I'll add a skip ci comment commit